webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/hooks/RunHooks.scala

```diff
@@ -323,44 +334,48 @@ object RunHooks {
-              r <- p.await.timeout(killTimeout).flatMap {
-                     case Some(ok) =>
-                       ok.succeed
-                     case None     =>
-                       val msg = s"Hook ${cmdInfo} timed out after ${killTimeout.render}"
-                       PureHooksLogger.LongExecLogger.error(msg) *> Unexpected(msg).fail
-                   }
-              _ <- f.interrupt
-              c  = translateReturnCode(path, r)
-              _ <- logReturnCode(c)
-            } yield {
-              history match {
-                case HookExecutionHistory.DoNotKeep => (c, Nil)
-                // Noop are not real hook execution, just filter them out from the result
-                case HookExecutionHistory.Keep      => (c, (x :: historyList).filterNot(_ == Noop))
+    val runAllSeq: IOResult[(HookReturnCode, List[(String, HookReturnCode)])] = {
+      ZIO.foldLeft(hooks.hooksFile)((Noop: HookReturnCode, List.empty[(String, HookReturnCode)])) {
+        case ((previousCode, historyList), (nextHookName, timeouts)) =>
+          val killTimeout = timeouts.kill.getOrElse(unitKillAfter)
+          val warnTimeout = timeouts.warn.getOrElse(unitWarnAfter)
+          previousCode match {
+            case x: Error   => (x, historyList).succeed
+            case x: Success => // run the next hook
+              val path    = hooks.basePath + File.separator + nextHookName
+              val env     = envVariables.add(hookParameters)
+              val cmdInfo = s"'${path}' with environment parameters: [${hookParameters.debugString}]"
+
+              for {
+                _ <- PureHooksLogger.For(logIdentifier).debug(s"Run hook: ${cmdInfo}")
+                _ <- PureHooksLogger.For(logIdentifier).trace(s"System environment variables: ${envVariables.debugString}")
+                f <- PureHooksLogger.LongExecLogger
+                       .warn(s"Hook is taking more than ${warnTimeout.render} to finish: ${cmdInfo}")
+                       .delay(warnTimeout)
+                       .fork
+                p <- RunNuCommand.run(Cmd(path, Nil, env.toMap, Some(hooks.basePath)))
+                r <- p.await.timeout(killTimeout).flatMap {
+                       case Some(ok) =>
+                         ok.succeed
+                       case None     =>
+                         val msg = s"Hook ${cmdInfo} timed out after ${killTimeout.render}"
+                         PureHooksLogger.LongExecLogger.error(msg) *> Unexpected(msg).fail
+                     }
+                _ <- f.interrupt
+                c  = translateReturnCode(path, r)
+                _ <- logReturnCode(c)
+              } yield {
+                history match {
+                  case HookExecutionHistory.DoNotKeep => (c, Nil)
+                  // Noop are not real hook execution, just filter them out from the result
+                  case HookExecutionHistory.Keep      => (c, ((nextHookName, c) :: historyList).filterNot(_._2 == Noop))
```

Comment:
we now keep the last code also as head of the history, because we need the name. 
So we could just have no first returned value and use a non empty list with some default value and check the head, but I think we will want to have different global value and last one at some point - perhaps in case of warning for example.